### PR TITLE
feat(authN) Implement OIDC

### PR DESCRIPTION
**Pull request by michalkrzyz**
Add oidc auth method
Add oidc mock service
Add e2e tests for OIDC auth
Add config variable to enable oidc
Remove duplicated Makefile targets

## Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Related Issue # (issue)
- Closes # (issue)
- Fixes # (issue)

> Remove if not applicable

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🤝 Documentation pages updated
- [x] 🙅 no documentation needed
- [ ] (if applicable) generated OpenAPI docs for CRD changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -84,15 +84,6 @@ test-app: gqlgen install-build-dependencies
 test-db: gqlgen install-build-dependencies
 	ginkgo -r internal/database/mariadb
 
-test-e2e: gqlgen install-build-dependencies
```
Comment (MR2011) [inline Makefile:87]: Probably a mistake?

Comment (michalkrzyz) [reply Makefile:87]: @MR2011 no it seems we have duplicated targets. Please check couple lines before.

File: internal/api/graphql/access/auth.go [hunk 2]
```diff
@@ -17,22 +17,23 @@ import (
 type Logger interface {
```
Comment (dorneanu) [inline internal/api/graphql/access/auth.go:17]: You don't actually need this interface. Why don't use just use
logrus.New() which gives you a new logger?

File: internal/api/graphql/access/no_auth_method.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+	"github.com/gin-gonic/gin"
+)
+
+const (
+	noAuthMethodName string = "NoAuthMethod"
+)
+
+type NoAuthMethod struct {
+	logger         Logger
```
Comment (dorneanu) [inline internal/api/graphql/access/no_auth_method.go:14]: Also here you don't need a custom interface. You can use
logrus.Logger
(https://sourcegraph.com/github.com/sirupsen/logrus/-/blob/logger.go?L17).

File: internal/api/graphql/access/oidc_auth_method.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	Exp           int    `json:"exp"`
+	IAT           int    `json:"iat"`
+	FirstName     string `json:"first_name"`
+	FamilyName    string `json:"family_name"`
+	JTI           string `json:"jti"`
+	Email         string `json:"email"`
+}
+
+type OidcAuthMethod struct {
+	logger   Logger
```
Comment (dorneanu) [inline internal/api/graphql/access/oidc_auth_method.go:36]: Use logrus.Logger

File: internal/api/graphql/access/oidc_auth_method.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+func (oam OidcAuthMethod) Verify(c *gin.Context) error {
+	rawToken, err := getAuthTokenFromHeader(oidcAuthHeader, c)
+	if err != nil {
+		return verifyError(oidcAuthMethodName, err)
+	}
+
+	token, err := oam.verifier.Verify(c.Request.Context(), rawToken)
+	if err != nil {
+		return verifyError(oidcAuthMethodName, err)
```
Comment (dorneanu) [inline internal/api/graphql/access/oidc_auth_method.go:73]: Here I guess you should check for different error types. It might be
the case that the token is not set OR the verification failed.

File: internal/api/graphql/access/oidc_auth_method.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+		oidcVerifier := oidcProvider.Verifier(oidcConfig)
+
+		l.Info("Initializing OIDC auth")
+		return &OidcAuthMethod{logger: l, provider: oidcProvider, config: oidcConfig, verifier: oidcVerifier}
+	}
+	return nil
+}
+
+func (oam OidcAuthMethod) Verify(c *gin.Context) error {
```
Comment (dorneanu) [inline internal/api/graphql/access/oidc_auth_method.go:65]: You should also verify the audience ID: https://www.ory.sh/docs/hydra/guides/audiences

File: internal/api/graphql/access/token_auth_method_test.go [hunk 7]
```diff
@@ -4,121 +4,57 @@
...
 	"github.com/cloudoperators/heureka/internal/util"
 )
 
 const (
-	testEndpoint    = "/testendpoint"
 	testScannerName = "testAccessScanner"
 	authTokenSecret = "xxx"
 )
 
-type noLogLogger struct {
```
Comment (dorneanu) [inline internal/api/graphql/access/token_auth_method_test.go:31]: What is this needed for?

**Review by dorneanu — CHANGES_REQUESTED** [review]

File: pkg/oidc/provider.go [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+	oidcProvider.initRsa()
+	oidcProvider.addRoutes()
+	return &oidcProvider
+}
+
+func (p *Provider) Start() {
+	p.ctx, p.cancel = context.WithCancel(context.Background())
+	p.server = &http.Server{Handler: p.router.Handler()}
+
+	serverAddr := "'default"
```
Comment (MR2011) [inline pkg/oidc/provider.go:50]: looks like a typo?

File: internal/api/graphql/access/test/server.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+const (
+	testEndpoint = "/testendpoint"
+)
+
+type TestServer struct {
+	port           string
+	auth           *access.Auth
+	cancel         context.CancelFunc
+	ctx            context.Context
```
Comment (dorneanu) [inline internal/api/graphql/access/test/server.go:24]: The way you use the context is not really clear to me. Usually you **propagate** a context through your requests. If I'd suggest a slight change: 

``` go
type TestServer struct {
	port           string
	auth           *access.Auth
	cancel         context.CancelFunc
	ctx            context.Context
	srv            *http.Server
	lastRequestCtx context.Context
}

func NewTestServer(auth *access.Auth) *TestServer {
	port := GetRandomFreePort() // Assuming this function is defined elsewhere
	return &TestServer{
		port: port,
		auth: auth,
	}
}

func (ts *TestServer) StartInBackground(ctx context.Context) error {
	ts.ctx, ts.cancel = context.WithCancel(ctx)
	r := gin.Default()
	r.Use(ts.auth.GetMiddleware())
	r.GET(testEndpoint, func(c *gin.Context) {
		c.JSON(http.StatusOK, gin.H{"message": "ok"})
	})

	ts.srv = &http.Server{Addr: fmt.Sprintf(":%s", ts.port), Handler: r}

	if err := util2.FirstListenThenServe(ts.srv); err != nil {
		return err
	}

	go func() {
		if err := ts.srv.ListenAndServe(); err != http.ErrServerClosed {
			fmt.Println("Server error:", err)
		}
	}()

	return nil
}

func (ts *TestServer) Stop(ctx context.Context) {
	fmt.Println("Shutting down the server...")
	ts.cancel()

	ctxTimeout, cancelTimeout := context.WithTimeout(ctx, 5*time.Second)
	defer cancelTimeout()

	if err := ts.srv.Shutdown(ctxTimeout); err != nil {
		fmt.Println("Server forced to shutdown:", err)
	}

	fmt.Println("Server exiting")
}
```

This way the caller (whoever runs the server / calls `StartInBackground`) has to provide the context. This also aligns more with Golang's philosophy.

File: internal/api/graphql/access/test/util.go [hunk 10]
```diff
@@ -49,6 +50,14 @@ func ExpectErrorMessage(resp *http.Response, expectedMsg string) {
...
 
+func ExpectRegexErrorMessage(resp *http.Response, expectedRegexMsg string, args ...interface{}) {
+	var respData struct {
+		Error string `json:"error" required:true`
+	}
+	unmarshalResponseData(resp, &respData)
+	Expect(respData.Error).Should(MatchRegexp(expectedRegexMsg, args...))
+}
+
 type Jwt struct {
```
Comment (dorneanu) [inline internal/api/graphql/access/test/util.go:52]: I'm still not convinced that important structures like `JWT` should be part of `util.go`. 
While at the moment you have following code structure: 

``` shell
├── access_test.go
├── auth.go
├── no_auth_method.go
├── oidc_auth_method.go
├── oidc_auth_method_test.go
├── test
│   ├── server.go
│   └── util.go
├── token_auth_method.go
├── token_auth_method_test.go
└── util.go
```

for better readability I'd suggest having something like:

``` shell
.
├── access
│   ├── access_test.go
│   ├── auth.go
│   ├── no/auth/method.go
│   ├── oidc/auth/method.go
│   ├── oidc/auth/method_test.go
│   ├── token/auth/method.go
│   ├── token/auth/method_test.go
│   └── util.go
├── oidc
│   ├── oidc_auth.go
│   ├── oidc/auth/test.go
│   └── util.go
└── util
    ├── server.go
    └── util.go
```

Where:
    
*Access Package*:
 - Contains the core functionality for authentication methods such as ~Jwt~, ~no_auth_method~, and
     ~token_auth_method~.

*OIDC Package*:
 - This package is specifically for handling OpenID Connect (OIDC) related functionalities.
 - It includes the ~oidc_auth.go~ file where you can define and manage OIDC-related structures and methods.
 - The existing files like ~oidc_auth_method.go~, ~oidc_auth_method_test.go~, and ~util.go~ will
     be moved to this package.

*Util Package*:
 - This package contains utility functions that are not specific to any particular authentication method but are still useful across the project.
 - It includes the existing files like ~server.go~.

File: internal/api/graphql/access/token_auth_method.go [hunk 11]
```diff
@@ -56,33 +47,20 @@ func (tam TokenAuthMethod) Verify(c *gin.Context) error {
...
-func getTokenFromHeader(c *gin.Context) (string, error) {
-	var err error
-	tokenString := c.GetHeader(tokenAuthHeader)
-	if tokenString == "" {
-		err = verifyError("No authorization header")
-	}
-	return tokenString, err
-}
-
 func (tam TokenAuthMethod) verifyTokenAndGetClaimsFromTokenString(tokenString string) (*TokenClaims, error) {
```
Comment (dorneanu) [inline internal/api/graphql/access/token_auth_method.go:77]: This purpose of this method is not really clear to be. Perhaps it's the naming. I propose renaming it to:

``` go
func (tam *TokenAuthMethod) parseTokenClaims(tokenString string) (*TokenClaims, error) {
	claims := &TokenClaims{}
	token, err := jwt.ParseWithClaims(tokenString, claims, tam.parse)
	if err != nil {
        ...
	}
	if !token.Valid {
        ...
	}
	return claims, nil
}

```

File: internal/api/graphql/access/token_auth_method_test.go [hunk 12]
```diff
@@ -4,121 +4,57 @@
...
+		testServer = NewTestServer(auth)
+		testServer.StartInBackground()
 	})
 
 	AfterEach(func() {
-		testServer.stop()
+		testServer.Stop()
 	})
 
 	When("Scanner access api through token auth middleware with valid token", func() {
```
Comment (dorneanu) [inline internal/api/graphql/access/token_auth_method_test.go:95]: You basically test if the token is valid or not. However, I'm missing some edge cases: 

- Generate a token with an expiration time in the past and verify that it is rejected
- Make a request without the ~X-Service-Authorization~ header and verify that it returns a 401 status code
- Generate a token with a different secret and verify that it is rejected
- Test cases where some of the claims are missing or invalid

**Review by dorneanu — CHANGES_REQUESTED** [review]
I've now had a deeper look at the code and made some observations which not necessarily address changes made in this PR. First of all the code structure is a little bit confusing. For better readbility I suggest refactoring the code alittle bit.

File: internal/api/graphql/access/oidc_auth_method.go (hunk 5 above)
Comment (michalkrzyz) [reply internal/api/graphql/access/oidc_auth_method.go:73]: What for?
The purpose of this return is creation of wrapped error message in oidc_auth_method->verify, so middleware sends back 401 (unauthorized) with error message in json.
Do we want to implement something else for some use case?

File: internal/api/graphql/access/oidc_auth_method.go (hunk 6 above)
Comment (michalkrzyz) [reply internal/api/graphql/access/oidc_auth_method.go:66]: Added test to prove that nothing has to be changed:
https://github.com/cloudoperators/heureka/blob/8ed3ada22937ec0e644776707658260aa59a3132/internal/e2e/oidc_auth_test.go#L115-L123

File: internal/api/graphql/access/no_auth_method.go (hunk 3 above)
Comment (michalkrzyz) [reply internal/api/graphql/access/no_auth_method.go]: Done

File: internal/api/graphql/access/auth.go (hunk 2 above)
Comment (michalkrzyz) [reply internal/api/graphql/access/auth.go:17]: Done

File: internal/api/graphql/access/token_auth_method_test.go (hunk 7 above)
Comment (michalkrzyz) [reply internal/api/graphql/access/token_auth_method_test.go:31]: Removed

File: internal/api/graphql/access/token_auth_method_test.go (hunk 12 above)
Comment (michalkrzyz) [reply internal/api/graphql/access/token_auth_method_test.go:95]: As they are rather high level of requirement for oidc they have been tested in e2e already.

1) https://github.com/cloudoperators/heureka/blob/8ed3ada22937ec0e644776707658260aa59a3132/internal/e2e/token_auth_test.go#L65C1-L72C4

2) https://github.com/cloudoperators/heureka/blob/8ed3ada22937ec0e644776707658260aa59a3132/internal/e2e/token_auth_test.go#L51C1-L57C4

3) https://github.com/cloudoperators/heureka/blob/8ed3ada22937ec0e644776707658260aa59a3132/internal/e2e/token_auth_test.go#L73C1-L80C4

4) tested invalid signing method and invalid token, but no other checks are done in the code, you can put anything in claims. Do we need to specify additional requirements for some claims?

File: internal/api/graphql/access/token_auth_method.go (hunk 11 above)
Comment (michalkrzyz) [reply internal/api/graphql/access/token_auth_method.go:77]: Changed to 'parseTokenWithClaims'

File: internal/api/graphql/access/oidc_auth_method.go (hunk 5 above)
Comment (dorneanu) [reply internal/api/graphql/access/oidc_auth_method.go]: Yeah, you're right. I just checked. You also wrap the error. Fine with me.

File: internal/api/graphql/access/test/util.go (hunk 10 above)
Comment (michalkrzyz) [reply internal/api/graphql/access/test/util.go:52]: Changed

File: internal/api/graphql/access/token_auth_method_test.go (hunk 12 above)
Comment (dorneanu) [reply internal/api/graphql/access/token_auth_method_test.go:95]: Ok, nice. 

> tested invalid signing method and invalid token, but no other checks are done in the code, you can put anything in claims. Do we need to specify additional requirements for some claims?

For now, it's ok.

**Review by dorneanu — APPROVED** [review]
Way better now. Thanks!

**Review by MR2011 — APPROVED** [review]
